### ISSUE TITLE
suggestion, don't allow store to dispatch action.type of undefined or null

### ISSUE THREAD

**Issue by kswope** [rackt/redux#541]
I've haven't given this a lot of deep though so I'll probably someday want this post erased from the internet, but a few times already I've screwed up my 'constants', 

You can see in the below example, if I misspell ACCRUE either here or in the constants file (I'm not even sure if its spelled correctly here), it becomes undefined and then the reducer happily uses the default switch, and because store.dispatch is inclined to put whatever it pleases (@@init/etc) through the default path, I can't use that to catch screw ups.  Why not throw an exception when a falsey action.type is dispatched, since its so easy to do and sometimes difficult to catch, even in tests.

```
     switch ( action.type ) {                                                                                                                        
       case constants.AUTH_TOKEN:                                                                                                                    
         return authenticate(state, action)                                                                                                          
       case constants.ACCRUE:                                                                                                                        
         return accrue(state, action)                                                                                                                
       default:                                                                                                                                      
         return state                                                                                                                                
     }   

```

**Comment by merk** [comment]
You can define your own middleware that would behave this way.

**Comment by danmartinez101** [comment]
Are there any cases where undefined would a valid action type? This
suggestion would help avoid a pretty annoying problem that is really fairly
trivial, but very easy to overlook, resulting in wasted time. The reason I
ask if there are cases where undefined would be valid is that I'd hate to
see use cases blocked by unnecessary constraints, but if there aren't
any/many then making this the default would be better than putting it in
middleware (which people probably won't remember to add until after they've
pulled out their hair). If not an error, at least a warning?

**Comment by gaearon** [comment]
I agree it's something we should do by default. This would also effectively force everyone to use "type" instead of bikeshedding on the property name, which is good for the ecosystem. Anybody with use cases for empty type?

**Comment by kswope** [comment]
Looking at this semantically, is an action.type of undefined or null even worthy of the term "action"?

At its worst being opinionated about falsey actions could discourage some smelly code.

**Comment by somebody32** [comment]
But what if there is some logic inside the action creator that decides that there is no need to proceed with dispatching (just think about any conditional logic that uses `getState`)?
Now you can "abort" dispatching, just by returning `{}` from the action creator (not really abort, just reducers will ignore this action entirely).
But if the action type will be required, workaround will be way more verbose: `{type: PLEASE_IGNORE_ME }`  + having `PLEASE_IGNORE_ME` in the constants file

**Comment by johanneslumpe** [comment]
@somebody32 that action creator could then just not call `dispatch`?

**Comment by wmertens** [comment]
But then you need to thunk, whereas a plain action creator can't just
return e.g. false right now.

On Sun, Aug 16, 2015, 11:37 Johannes Lumpe <notifications@github.com> wrote:

> @somebody32 <https://github.com/somebody32> that action creator could
> then just not call dispatch?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rackt/redux/issues/541#issuecomment-131512659>.
>
-- 

Wout.
(typed on mobile, excuse terseness)

**Comment by gaearon** [comment]
>But what if there is some logic inside the action creator that decides that there is no need to proceed with dispatching (just think about any conditional logic that uses getState)?

As @wmertens notes, you can do that just fine with `redux-thunk`. It doesn't *force* you to call `dispatch()`. You can `return;` from the returned function, and it will just work.

**Comment by wmertens** [comment]
Still, I kinda feel like a plain action creator should be allowed to return
false (exactly that, nothing else) to not dispatch anything, much like
React allows to not create an element when you return null or false.

Obviously, that's trivial middleware, but still…

On Mon, Aug 17, 2015, 17:43 Dan Abramov <notifications@github.com> wrote:

> But what if there is some logic inside the action creator that decides
> that there is no need to proceed with dispatching (just think about any
> conditional logic that uses getState)?
>
> As @wmertens <https://github.com/wmertens> notes, you can do that just
> fine with redux-thunk. It doesn't *force* you to call dispatch(). You can
> return; from the returned function, and it will just work.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rackt/redux/issues/541#issuecomment-131867753>.
>
-- 

Wout.
(typed on mobile, excuse terseness)

**Comment by gaearon** [comment]
>Still, I kinda feel like a plain action creator should be allowed to return
false (exactly that, nothing else) to not dispatch anything, much like
React allows to not create an element when you return null or false.

I don't agree. Any real use case I have in mind involves `getState()` which a plain action creator doesn't have access to anyway.

**Comment by wmertens** [comment]
D'oh you're absolutely right...
and of course actioncreators could also send an action of type "ERROR" or
whatever.

On Mon, Aug 17, 2015 at 9:04 PM Dan Abramov <notifications@github.com>
wrote:

> Still, I kinda feel like a plain action creator should be allowed to return
> false (exactly that, nothing else) to not dispatch anything, much like
> React allows to not create an element when you return null or false.
>
> I don't agree. Any real use case I have in mind involves getState() which
> a plain action creator doesn't have access to anyway.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rackt/redux/issues/541#issuecomment-131931404>.
>
-- 

Wout.
(typed on mobile, excuse terseness)

**Comment by gaearon** [comment]
Done in 3.0.0.

**Comment by frankoid** [comment]
Is this change compatible with react-thunk? I'm trying to upgrade https://github.com/frankoid/flux-comparison to redux 3, but when I [upgrade from 2.0.0 to 3.0.2](https://github.com/frankoid/flux-comparison/commit/01f53424be4110e39c11a78e93eaffab18829be2) then I start seeing an `Actions may not have an undefined "type" property. Have you misspelled a constant?` error.

**Comment by gaearon** [comment]
It's compatible. We're actually doing a good thing in 3.0: [this line](https://github.com/voronianski/flux-comparison/blob/d6675d76544710b7bd222eb6f80c36552ae9d0c9/redux/js/actions/ActionCreators.js#L7) uses a constant [that was never defined](https://github.com/voronianski/flux-comparison/blob/d6675d76544710b7bd222eb6f80c36552ae9d0c9/redux/js/constants/ActionTypes.js#L3-L8).

**Comment by gaearon** [comment]
@frankoid By the way, if you don't mind, I'd like taking a stab at updating it myself. There are a few things I'd change there..

**Comment by frankoid** [comment]
@gaearon thanks! I've got it working now. Happy for you to update it yourself - I am learning React and Redux and was using the upgrade as a learning exercise, so I'm not up to speed with Redux best practices.

**Comment by gaearon** [comment]
@frankoid FYI, I updated it in the flux-comparison repo.

**Comment by Sigfried** [comment]
I'm still new to all this and probably doing a wrong thing, but my action creator looks like:

```
export const setPathname = pathname => {
  if (pathname === myrouter.history.location,pathname)
    return
  return reduxPush(myrouter.history.createHref({...myrouter.history.location,pathname}))
}
```

I tried returning {} for no change and that also complained, I tried {type:null}, but that still gets dispatched.

So, is there a right way to have an action creator (bound with bindActionCreators) that sometimes doesn't dispatch? Or should my use case (wanting to centralize all router stuff because I keep changing my mind about how it should work) just be handled some totally other way?

Sorry to be asking here instead of SO, but this seems to be where the relevant discussion is.
